### ISSUE TITLE
EKS: autoscaler: allow scaleout when adding pods would exceed number of available IPs on a node

### ISSUE THREAD

**Issue by ps2goat** [aws/containers-roadmap#1267]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
I ran into similar errors today as these issues:
- https://github.com/kubernetes/autoscaler/issues/2210
- https://github.com/aws/amazon-vpc-cni-k8s/issues/1245
TLDR:
> (combined from similar events): Failed create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "redacted" network for pod "redacted": NetworkPlugin cni failed to set up pod "redacted" network: add cmd: failed to assign an IP address to container

The suggestion there was to make an issue here for a feature request. I did not find one by searching all issues for "autoscaler ip", hence this request.

It seems like the error in these threads and in my EKS cluster is due to a pod being deployed onto a node that has no more network interfaces available on it to support the pods.

**Which service(s) is this request for?**
EKS

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
I made 3 deployments, each with 1 pod, via our CI/CD pipeline, and one out of 3 came back with this error. It is the first time I've seen it in our setup in 2+ years across 3 clusters. I'd like to see the autoscaler/balancer (or whatever is involved) have improved reliability.

A redeploy works, but before that, the balancer/scaler didn't recognize that this one pod needed to be put on a different node. It was stuck in the above error state for 2+ hours before I saw the issue. (I'm currently in a non-production environment!)

**Are you currently working around this issue?**
A redeploy worked.

**Additional context**
These 3 deployments get deployed together for reasons. So one Gitlab pipeline deploys all 3, one at a time, but in succession as part of the same operation. kubectl is used to create the deployments within kubernetes.

We're currently on platform version eks.12, kubernetes version 1.14 (Yes, we need to update it-- if this was fixed, please let me know what version!)

**Comment by rohith-gittest** [comment]
EKS worker nodes has limitations  for running number of pod based on the instance type. In our case we use instance type-c5.2xlarge which allows 58 pods per instance. If the pods are scheduled more than that number on the same instance, then those pods are getting stuck in ContainerCreating status.
Describing the pods gives us following errors:  " Failed to create pod sandbox: rpc error: code = Unknown desc = failed to set up sandbox container "8be3026cc6bb2b311570f118dfef8ab93ae491e6d6fc20e12a46e9b814cff716" network for pod "pod-name": networkPlugin cni failed to set up pod "pod-name" network: add cmd: failed to assign an IP address to container "

As part of solution we tried adding --use-max-pods false --kubelet-extra-args '--max-pods=110' into our EKS worker node bootstrap script. Still the issue remains same.
Please guide us the solutions if there are any to remove pods restriction based on instance type and run max number of pods in a worker node.

**Comment by snolan-amount** [comment]
Note that this has changed considerably in CNI 1.9 - you can now have many more pods per worker node - in fact AWS now recommend limiting the number of pods to less than the available IP addresses because of Kubernetes limits which recommend no more than 110 pods per node (https://kubernetes.io/docs/setup/best-practices/cluster-large/). This is made possible because IPs are now allocated in /28 ranges rather than individually. So use CNI custom networking (CNI 1.9 or higher) and you can get 110 pods on most instance types (more if you're happy to exceed the K8s recommendations).
See the ENABLE_PREFIX_DELEGATION env var and 
https://docs.aws.amazon.com/eks/latest/userguide/cni-increase-ip-addresses.html

**Comment by bmihaescu** [comment]
Hi @rohith-mr-rao, did you manage to fix your problem in some way? We experience the same issue and I'm curious if you found a workaround.
